### DISCUSSION TITLE
Question on Explicit Output Dimensions

### DISCUSSION THREAD

**Discussion by nukelash** [onnx/onnx#4604] [Q&A]
Hi, I am fairly new to Onnx and am trying to use the C++ onnxruntime to run inference on a model trained in pytorch. I've been running into issues with dimensions of -1 in my output tensors, and while investigating that, noticed the dimensions of my output on Netron aren't "explicit". In other words, all of my input nodes show something like "name: input_spec type: float32[1,1,3,241,2]" whereas my output nodes say "name: output_spec type: float32[ScatterNDoutput_spec_dim_0,ScatterNDoutput_spec_dim_1,ScatterNDoutput_spec_dim_2,ScatterNDoutput_spec_dim_3,ScatterNDoutput_spec_dim_4]" (ScatterND is the node right before the output node). However when I inspect other onnx models such as resnet, these output dimensions are explicit (save for batch size N)

My model was not exported with dynamic axis, and as far as I understand, it shouldn't because all of my inferences will be single batch, fixed sized input, and fixed sized output.

My questions are:
- What would be the cause of having output dimensions not explicitly shown in Netron? (something in my model architecture, how I exported it, etc)
- How can I change it so that these values are explicit?
- Does anyone know if this is actually the cause of -1 dimensions being returned by GetShape() in c++ onnxruntime?

Thanks in advance!

**Comment by zhenhuaw-me** [comment]
Hi @nukelash,

> What would be the cause of having output dimensions not explicitly shown in Netron?

It's the exporter which generates the ONNX model, e.g. PyTorch. Sometimes, the exporter just doesn't try its best to figure out the shapes of each tensor. But don't worry, each ONNX importer can take care of that.

> How can I change it so that these values are explicit?

You can try `onnxcli infershape /path/to/input/model.onnx /path/to/output/model.onnx`. See more here: https://github.com/zhenhuaw-me/onnxcli#infershape

> Does anyone know if this is actually the cause of -1 dimensions being returned by GetShape() in c++ onnxruntime?

Not sure how you exported the model, but anyway, it's the exporter.
